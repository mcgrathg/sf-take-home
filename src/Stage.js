import React from 'react';
import PropTypes from 'prop-types';
import { Spring } from 'react-spring/renderprops';
import styled from 'styled-components';
import StageHandler from './StageSelectionHandler';

const Highlighted = styled.span`
  position: relative;
  font-weight: 800;

  &::after {
    position: absolute;
    left: 0;
    bottom: 0;
    height: 0.5rem;
    width: ${props => props.width};
    content: '';
    background-color: #ccf2f0;
    z-index: -1;
  }
`;

const Stage = ({ onSubmit, options, prompt, highlight, promptEnd }) => {
  return (
    <Spring
      from={{ opacity: 0 }}
      to={{ opacity: 1 }}
      config={{ duration: 500 }}
    >
      {props => (
        <div style={props}>
          <p className="prompt">
            {prompt}
            <Spring
              from={{ number: 0 }}
              to={{ number: 100 }}
              delay={500}
              config={{ duration: 500 }}
            >
              {({ number }) => (
                <Highlighted width={`${number}%`}>{highlight}</Highlighted>
              )}
            </Spring>
            {promptEnd}
          </p>

          <StageHandler>
            {({ onOptionToggled, selected, resetSelections }) => {
              const optionToggles = options.map(option => {
                const isSelected = selected[option];

                return (
                  <button
                    key={option}
                    type="button"
                    aria-pressed={isSelected ? 'true' : 'false'}
                    className={`option ${isSelected ? 'checked' : ''}`}
                    onClick={() => onOptionToggled(option, !isSelected)}
                  >
                    {option}
                  </button>
                );
              });

              return (
                <>
                  <div className="options">{optionToggles}</div>
                  <button
                    className="cta"
                    type="button"
                    onClick={() => {
                      onSubmit(Object.keys(selected));
                      resetSelections();
                    }}
                  >
                    â†’
                  </button>
                </>
              );
            }}
          </StageHandler>
        </div>
      )}
    </Spring>
  );
};

Stage.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  prompt: PropTypes.string.isRequired,
  highlight: PropTypes.string.isRequired,
  promptEnd: PropTypes.string,
};

Stage.defaultProps = {
  promptEnd: '?',
};

export default Stage;
