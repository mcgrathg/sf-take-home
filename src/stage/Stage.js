import React from 'react';
import PropTypes from 'prop-types';
import FadeIn from './FadeIn';
import StageSelectionHandler from './StageSelectionHandler';
import Prompt from './Prompt';
import OptionToggles from './OptionToggles';
import SubmitSelectionsButton from './SubmitSelectionsButton';

const Stage = ({ onSubmit, options, prompt, highlight, className }) => (
  <FadeIn>
    <Prompt prompt={prompt} highlight={highlight} />

    <StageSelectionHandler>
      {({ onOptionToggled, selected, resetSelections, ...rest }) => {
        return (
          <>
            <OptionToggles
              options={options}
              onOptionToggled={onOptionToggled}
              selected={selected}
              className={className}
            />

            <SubmitSelectionsButton
              onClickHandler={() => {
                onSubmit(
                  Object.entries(selected).reduce((acc, [key, value]) => {
                    if (value) {
                      acc.push(key);
                    }
                    return acc;
                  }, []),
                );
                resetSelections();
              }}
            />
          </>
        );
      }}
    </StageSelectionHandler>
  </FadeIn>
);

Stage.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  prompt: PropTypes.string.isRequired,
  highlight: PropTypes.string.isRequired,
  className: PropTypes.string,
};

Stage.defaultProps = {
  className: '',
};

export default Stage;
