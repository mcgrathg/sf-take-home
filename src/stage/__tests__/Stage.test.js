import React from 'react';
import { render } from 'react-testing-library';
import { noOp } from 'test-utils';
import Stage from '../Stage';

test('renders', () => {
  const { container } = render(
    <Stage onSubmit={noOp} options={[]} prompt="" highlight="" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});
