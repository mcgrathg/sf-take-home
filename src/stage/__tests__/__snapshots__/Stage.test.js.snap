// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stage renders 1`] = `
.c0 {
  margin: 0 auto 1rem;
  color: #2f3237;
  font-size: 1.5rem;
  line-height: 1.75;
  font-weight: normal;
}

.c1 {
  position: relative;
  font-weight: 800;
}

.c1::after {
  position: absolute;
  left: 0;
  bottom: 0;
  height: 0.5rem;
  width: 100%;
  content: '';
  background-color: #ccf2f0;
  z-index: -1;
  -webkit-transition: all 0.5s;
  transition: all 0.5s;
  -webkit-transform-origin: left top;
  -ms-transform-origin: left top;
  transform-origin: left top;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
}

.c1.mounted::after {
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c2 {
  margin: 3rem 0;
}

.c3 {
  width: 13rem;
  max-width: 100%;
  height: 3.25rem;
  border: 1px solid #d43f43;
  padding: 0.5rem 1rem;
  color: #d43f43;
  background-color: #fff;
  -webkit-transition: all 0.25s ease-in;
  transition: all 0.25s ease-in;
}

.c3:hover,
.c3:focus {
  color: #fff;
  background-color: #d43f43;
}

.c3:active {
  -webkit-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
  -webkit-transition: none;
  transition: none;
}

.c3:focus {
  box-shadow: 0 0 0 3px #f2f0ee,0 0 0 5px #d43f43;
}

<div
  style="opacity: 0;"
>
  <h1
    class="c0"
  >
    
    <span
      class="c1"
    />
    ?
  </h1>
  <div
    class="c2"
  />
  <button
    class="c3"
    title="Next"
    type="button"
  >
    â†’
  </button>
</div>
`;
